NIFI-8644: Introduced a notion of ParameterProviderDefinition. Refact…

markap14 opened the pull request:
…ored stateless to use this when creating a dataflow so that Parameter Provider implementations can be externalized into NARs. Also updated ExtensionDiscoveringManager such that callers are able to provide a new type of class to be discovered (e.g., ParameterProvider) so that the extensions will be automatically discovered.

<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

gresockj commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/config/PropertiesFileFlowDefinitionParser.java:
```diff
@@ -174,6 +180,80 @@
+
+            if (relativePropertyName.startsWith("properties.")) {
+                if (relativePropertyName.length() < 12) {
+                    logger.warn("Encountered unexpected property <" + propertyName + "> in flow definition. This property will be ignored.");
```
If the user provides only an unexpected property and no other configuration, the definition has still already been added to the map on line 196.  Perhaps we can detect if the providers are valid before returning from this method, and if only unexpected properties were detected, don't add the definition.

markap14 replied:
Well, I don't think we know at this point whether or not it's valid. At this point, we are only parsing the definition. Validity would really depend on the implementation of the Parameter Provider.

gresockj commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/parameter/ParameterOverrideProvider.java:
```diff
@@ -15,19 +15,46 @@
+            final String propertyValue = entry.getValue();
+
+            final ParameterOverride override;
+            if (propertyName.contains(":")) {
```
Constant?

gresockj reviewed (changes requested):
Neat stuff.  I'd like to test it out using my own simple ParameterProvider, but this is my first pass of code review.  So far just some minor error handling suggestions.

gresockj commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/parameter/CompositeParameterProvider.java:
```diff
@@ -15,14 +15,12 @@
-public class CompositeParameterProvider implements ParameterProvider {
+public class CompositeParameterProvider extends AbstractParameterProvider implements ParameterProvider {
     private final List<ParameterProvider> parameterProviders;
 
```
After deploying a custom parameter provider, I realized that you probably need an additional `provider.isParameterDefined(contextName, parameterName);` at line 33 before you get the value.  What I noticed is that because my custom ParameterProvider returned false for isParameterDefined(), but the ParameterOverrideProvider returned true, the StandardStatelessEngine then proceeds to call `CompositeParameterProvider.getParameterValue(...)`.  Then, since the custom provider is first in the list, CompositeParameterProvider:33 gets the value from my custom provider (even though it returned false for `isParameterDefined(...)`) instead of from ParameterOverrideProvider.  I would expect the value to be taken from the latter.

gresockj reviewed (changes requested):
A custom parameter provider worked like a charm!  This is a really neat addition to the framework.  One minor issue in handling parameter provider failover.

gresockj commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/engine/StandardParameterProviderInitializationContext.java:
```diff
@@ -0,0 +1,49 @@
+    private static Map<PropertyDescriptor, String> createPropertyMap(final ParameterProvider provider, final Map<String, String> propertyValues) {
+        final Map<PropertyDescriptor, String> propertyMap = new LinkedHashMap<>();
+        for (final Map.Entry<String, String> entry : propertyValues.entrySet()) {
+            final PropertyDescriptor descriptor = provider.getPropertyDescriptor(entry.getKey());
```
Do you want to add parameter descriptor validation here?  In my custom parameter provider I specified that one of my property descriptors was required, and this requirement was not honored.

markap14 replied:
Whoops! It looks like we're not validating Reporting Tasks at the moment in stateless either! Will update the engine to ensure that they are valid. If not, I think we need to throw an IllegalStateException.

markap14 commented on nifi-stateless/nifi-stateless-assembly/README.md:
```diff
@@ -427,28 +427,29 @@ Additionally, there may be sensitive parameters that users prefer not to include
 Environment Variables, for example.
 
```
Whoops! Totally forgot that! Great catch!

markap14 commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/config/PropertiesFileFlowDefinitionParser.java:
```diff
@@ -174,6 +180,80 @@
+                }
+
+                final String providerPropertyName = relativePropertyName.substring(11);
+                definition.getPropertyValues().put(providerPropertyName, propertyValue);
```
Thanks. Made it really easy to replicate. I also caught a couple of other minor tweaks that should be added to clarify when things aren't as expected. Will update these.

markap14 commented:
Thanks for the thorough review @gresockj! I addressed the concerns and a few other minor tweaks that i found when updating the code. I also added a couple of system tests to verify that the behavior is as expected.

gresockj commented on nifi-stateless/nifi-stateless-assembly/README.md:
```diff
@@ -427,33 +431,54 @@ Additionally, there may be sensitive parameters that users prefer not to include
\ No newline at end of file
+will not be addressable.
+
+At times, none of the builtin capabilities for resolving Parameters are ideal, though. In these situations, we can use a custom Parameter Provider in order to source Parameter values from elsewhere.
```
I'd add a dash to "built-in".

gresockj reviewed (changes requested):
Nice, now the correct overriding logic is working.  It also looks like validation is kicking in now, though I encountered one issue with the validation, mentioned below.

gresockj commented on nifi-stateless/nifi-stateless-bundle/nifi-stateless-engine/src/main/java/org/apache/nifi/stateless/engine/StandardStatelessEngine.java:
```diff
@@ -182,6 +197,98 @@ public StatelessDataflow createFlow(final DataflowDefinition<VersionedFlowSnapsh
+            final String propertyValue = property.getValue();
+
+            final PropertyDescriptor descriptor = component.getPropertyDescriptor(propertyName);
+            final PropertyConfiguration propertyConfiguration = new PropertyConfiguration(propertyValue, null, Collections.emptyList());
```
Nice, that appears to have fixed it.  However.. one more problem when I try to override the parameter from the command line:

```
Exception in thread "main" java.lang.IllegalStateException: Could not create Parameter Provider Parameter Override Provider of type org.apache.nifi.stateless.parameter.ParameterOverrideProvider
	at org.apache.nifi.stateless.engine.StandardStatelessEngine.createParameterProvider(StandardStatelessEngine.java:245)
	at org.apache.nifi.stateless.engine.StandardStatelessEngine.createParameterProvider(StandardStatelessEngine.java:204)
	at org.apache.nifi.stateless.engine.StandardStatelessEngine.createFlow(StandardStatelessEngine.java:175)
	at org.apache.nifi.stateless.flow.StandardStatelessDataflowFactory.createDataflow(StandardStatelessDataflowFactory.java:223)
	at org.apache.nifi.stateless.bootstrap.StatelessBootstrap.createDataflow(StatelessBootstrap.java:66)
	at org.apache.nifi.stateless.bootstrap.RunStatelessFlow.createDataflow(RunStatelessFlow.java:94)
	at org.apache.nifi.stateless.bootstrap.RunStatelessFlow.main(RunStatelessFlow.java:56)
Caused by: java.lang.IllegalStateException: Parameter Provider with name <Parameter Override Provider> is not valid: ['Context:param' validated against 'joe' is invalid because 'Context:param' is not a supported property or has no Validator associated with it]
	at org.apache.nifi.stateless.engine.StandardStatelessEngine.createParameterProvider(StandardStatelessEngine.java:240)
	... 6 more
```

markap14 replied:
Ah yes, now that I have fixed the validation, I need that provider to actually be valid :( Will fix.

markap14 commented on nifi-system-tests/nifi-system-test-extensions-bundle/nifi-system-test-extensions/src/main/resources/META-INF/services/org.apache.nifi.stateless.parameter.ParameterProvider:
```diff
@@ -0,0 +1,18 @@
+
+org.apache.nifi.stateless.parameters.InvalidParameterProvider
+org.apache.nifi.stateless.parameters.NumericParameterProvider
+org.apache.nifi.stateless.parameters.ParameterProviderWithProperties
```
Yeah, so this is intended to be used only for tests. But if it were put into the src/test, it wouldn't get packaged into the jar. They are for testing but not for testing in this module.

gresockj replied:
Got it, forgot this was inside the nifi-system-test-extensions module already.

gresockj reviewed (commented):
Nice work, @markap14, this will be a useful addition to Stateless!  Everything appears to be working as designed now.  One last observation -- I notice that the custom parameter providers take precedence over the command line parameter overrides.  I would find it more intuitive to be able to override a custom provider using the command line param overrides, otherwise adding a custom provider effectively negates the command line override functionality.  What do you think?

markap14 commented:
Agreed @gresockj the command line override should take precedence. Followed by environment variables, and then anything provided by a ParameterProvider. Updated the PR to use this precedence. Thanks again!

exceptionfactory commented on nifi-system-tests/nifi-stateless-system-test-suite/src/test/java/org/apache/nifi/stateless/parameters/ParameterContextIT.java:
```diff
@@ -47,6 +50,174 @@
+        parameterContext.getParameters().add(createVersionedParameter("three", "-1"));  // Set value to -1. This should be overridden by the Numeric Parameter Context.
+        flowBuilder.getRootGroup().setParameterContextName("Context 1");
+
+        try {
+            loadDataflow(flowSnapshot, Collections.emptyList(), parameterProviders, Collections.emptySet(), TransactionThresholds.SINGLE_FLOWFILE);
+            Assert.fail("Expected to fail on startup because parameter provider is not valid");
+        } catch (final IllegalStateException expected) {
+        }
```
Is `loadDataflow()` expected to throw an IllegalStateException here?  If so, could `assertThrows()` be used instead of the try-catch and `Assert.fail()`?

markap14 replied:
Didn't realize that `assertThrows` had been added to JUnit 4. Will update that.

gresockj reviewed (approved):
Ok, runtime behavior looks good now!

exceptionfactory reviewed (approved):
Thanks for the update @markap14, and thanks for the thorough review @gresockj! +1 Merging.
